distribution/tools/entitlement-agent/impl/src/test/java/org/elasticsearch/entitlement/instrumentation/impl/InstrumenterTests.java

```diff
@@ -0,0 +1,151 @@
+ */
+
+package org.elasticsearch.entitlement.instrumentation.impl;
+
+import org.elasticsearch.entitlement.api.EntitlementChecks;
+import org.elasticsearch.entitlement.api.EntitlementProvider;
+import org.elasticsearch.entitlement.instrumentation.InstrumentationService;
+import org.elasticsearch.entitlement.instrumentation.MethodKey;
+import org.elasticsearch.logging.LogManager;
+import org.elasticsearch.logging.Logger;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Before;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Map;
+
+import static org.elasticsearch.entitlement.instrumentation.impl.ASMUtils.bytecode2text;
+
+/**
+ * This tests {@link InstrumenterImpl} in isolation, without a java agent.
+ * It causes the methods to be instrumented, and verifies that the instrumentation is called as expected.
+ * Problems with bytecode generation are easier to debug this way than in the context of an agent.
+ */
+@ESTestCase.WithoutSecurityManager
+public class InstrumenterTests extends ESTestCase {
+    final InstrumentationService instrumentationService = new InstrumentationServiceImpl();
+
+    private static TestEntitlementManager getTestChecks() {
+        return (TestEntitlementManager) EntitlementProvider.checks();
+    }
+
+    @Before
+    public void initialize() {
+        getTestChecks().isActive = false;
+    }
+
+    /**
+     * Contains all the virtual methods from {@link ClassToInstrument},
+     * allowing this test to call them on the dynamically loaded instrumented class.
+     */
+    public interface Testable {}
+
+    /**
+     * This is a placeholder for real class library methods.
+     * Without the java agent, we can't instrument the real methods, so we instrument this instead.
+     * <p>
+     * Methods of this class must have the same signature and the same static/virtual condition as the corresponding real method.
+     * They should assert that the arguments came through correctly.
+     * They must not throw {@link TestException}.
+     */
+    public static class ClassToInstrument implements Testable {
+        public static void systemExit(int status) {
+            assertEquals(123, status);
+        }
+    }
+
+    static final class TestException extends RuntimeException {}
+
+    /**
+     * We're not testing the permission checking logic here.
+     * This is a trivial implementation of {@link EntitlementChecks} that just always throws,
+     * just to demonstrate that the injected bytecodes succeed in calling these methods.
+     */
+    public static class TestEntitlementManager implements EntitlementChecks {
+        /**
+         * This allows us to test that the instrumentation is correct in both cases:
+         * if the check throws, and if it doesn't.
+         */
+        volatile boolean isActive;
+
+        @Override
+        public void checkSystemExit(Class<?> callerClass, int status) {
+            throwIfActive();
```

Comment:
could we validate the callerClass? And also the status (to make sure it is passed through correctly)?

Reply:
Great idea!